# Pandas Review using a Pokemon Dataset

In [53]:
import pandas as pd

df = pd.read_csv('https://gist.githubusercontent.com/armgilles/\
194bcff35001e7eb53a2a8b441e8b2c6/raw/92200bc0a673d5ce2110aaad4544ed6c4010f687/pokemon.csv')

### Practice Exercises

#### 1. Obtain the following information:
    
    - dimensions
    - dtypes
    - column names
    - summary statistics


In [54]:
df.shape

(800, 13)

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [56]:
df.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [57]:
df.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

#### 2. Change all the column names
      - lowercase letters
      - remove whitespace
      - remove periods

In [58]:
df.columns = df.columns.str.lower()
df =df.rename(columns={'#':'number'})
df =df.rename(columns={'type 1':'type_1'})
df =df.rename(columns={'type 2':'type_2'})
df =df.rename(columns={'sp. atk':'sp_atk'})
df =df.rename(columns={'sp. def':'sp_def'})


df.columns


Index(['number', 'name', 'type_1', 'type_2', 'total', 'hp', 'attack',
       'defense', 'sp_atk', 'sp_def', 'speed', 'generation', 'legendary'],
      dtype='object')

#### 3. What is the highest HP value present?

A: 255

In [59]:
df.hp.max()

255

#### 4. Which Pokemon possess(es) the highest HP value?
- note: the `name` column is the pokemon name

A: Blissey

In [60]:
hp_series = df['hp'] == df['hp'].max()
highest_hp = df[hp_series]
highest_hp

number     name  type_1 type_2  total   hp  attack  defense  sp_atk  \
261     242  Blissey  Normal    NaN    540  255      10       10      75   

     sp_def  speed  generation  legendary  
261     135     55           2      False

#### 5. How many different types are represented in Type 1? What are they?


A: 18,  ['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying']

In [61]:
print(df.type_1.nunique())
df.type_1.unique()

18


array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

#### 6. Number of Pokemon whose Type 2 is Ghost

In [62]:
ghost_series = df['type_2'] == 'Ghost'
type_2_ghost = df[ghost_series]
len(type_2_ghost)


14

A: 14

#### 7. Percentage of Pokemon whose Type 2 is Ghost

In [63]:
round((len(type_2_ghost)/len(df.type_2))*100, 2)

1.75

A: 1.75

#### 8. Number of Pokemon whose Attack is greater than Defense

In [64]:
great_attack = (df.attack)>(df.defense)
great_attack.sum()

433

A: 433

#### 9. What is the slowest speed for Grass type (either type 1 or type 2)

A: 10

In [65]:
df.speed.min()


5

In [45]:
grss_series = df['type_1'] == 'Ghost'
type_2_ghost = df[ghost_series]
type_2_ghost.speed.min()



20

In [46]:
grss_seriess = df['type_1'] == 'Ghost'
type_1_ghost = df[grss_seriess]
type_1_ghost.speed.min()

nan

In [ ]:
grs_series

In [42]:
grass_series=df['type_1']='Grass'
grass_series.speed.min()

# grass_poke= grass_series.min("speed")




AttributeError: 'str' object has no attribute 'speed'

In [39]:
grass_pokemon = df[df['type_1'] == 'Grass']
grass_pokemon['speed'].min()

# Find the slowest speed
# slowest_speed = grass_pokemon['speed'].min()
# slowest_speed

5

#### 10. Find the average speed by Generation

A: generation
- 1:    72.6
- 2:    61.8
- 3:    66.9
- 4:    71.3
- 5:    68.1
- 6:    66.4

In [47]:
df.groupby('generation').speed.mean()


generation
1    72.584337
2    61.811321
3    66.925000
4    71.338843
5    68.078788
6    66.439024
Name: speed, dtype: float64

#### 11. How many LEGENDARY Pokemon are DRAGON type?

A: 16

In [93]:
df3=df[(df['legendary']==True) & (df['type_1']== 'Dragon') |(df['type_2']== 'Dragon')]
df3

number                       name    type_1    type_2  total   hp  \
7         6  CharizardMega Charizard X      Fire    Dragon    634   78   
196     181      AmpharosMega Ampharos  Electric    Dragon    610   90   
249     230                    Kingdra     Water    Dragon    540   75   
275     254      SceptileMega Sceptile     Grass    Dragon    630   70   
360     329                    Vibrava    Ground    Dragon    340   50   
361     330                     Flygon    Ground    Dragon    520   80   
417     380                     Latias    Dragon   Psychic    600   80   
418     380          LatiasMega Latias    Dragon   Psychic    700   80   
419     381                     Latios    Dragon   Psychic    600   80   
420     381          LatiosMega Latios    Dragon   Psychic    700   80   
425     384                   Rayquaza    Dragon    Flying    680  105   
426     384      RayquazaMega Rayquaza    Dragon    Flying    780  105   
540     483                     Dialga     Steel    Dragon    680  100   
541     484                     Palkia     Water    Dragon    680   90   
544     487      GiratinaAltered Forme     Ghost    Dragon    680  150   
545     487       GiratinaOrigin Forme     Ghost    Dragon    680  150   
694     633                      Deino      Dark    Dragon    300   52   
695     634                   Zweilous      Dark    Dragon    420   72   
696     635                  Hydreigon      Dark    Dragon    600   92   
706     643                   Reshiram    Dragon      Fire    680  100   
707     644                     Zekrom    Dragon  Electric    680  100   
710     646                     Kyurem    Dragon       Ice    660  125   
711     646         KyuremBlack Kyurem    Dragon       Ice    700  125   
712     646         KyuremWhite Kyurem    Dragon       Ice    700  125   
761     691                   Dragalge    Poison    Dragon    494   65   
766     696                     Tyrunt      Rock    Dragon    362   58   
767     697                  Tyrantrum      Rock    Dragon    521   82   
790     714                     Noibat    Flying    Dragon    245   40   
791     715                    Noivern    Flying    Dragon    535   85   
794     718           Zygarde50% Forme    Dragon    Ground    600  108   

     attack  defense  sp_atk  sp_def  speed  generation  legendary  
7       130      111     130      85    100           1      False  
196      95      105     165     110     45           2      False  
249      95       95      95      95     85           2      False  
275     110       75     145      85    145           3      False  
360      70       50      50      50     70           3      False  
361     100       80      80      80    100           3      False  
417      80       90     110     130    110           3       True  
418     100      120     140     150    110           3       True  
419      90       80     130     110    110           3       True  
420     130      100     160     120    110           3       True  
425     150       90     150      90     95           3       True  
426     180      100     180     100    115           3       True  
540     120      120     150     100     90           4       True  
541     120      100     150     120    100           4       True  
544     100      120     100     120     90           4       True  
545     120      100     120     100     90           4       True  
694      65       50      45      50     38           5      False  
695      85       70      65      70     58           5      False  
696     105       90     125      90     98           5      False  
706     120      100     150     120     90           5       True  
707     150      120     120     100     90           5       True  
710     130       90     130      90     95           5       True  
711     170      100     120      90     95           5       True  
712     120       90     170     100     95           